tokio-rustls/src/common/mod.rs

```diff
@@ -288,12 +286,7 @@ where
 
                 Err(err) if err.kind() == io::ErrorKind::UnexpectedEof => {
                     self.eof = true;
-                    self.unexpected_eof = true;
-                    if prev == buf.remaining() {
-                        Poll::Ready(Err(err))
-                    } else {
-                        break;
-                    }
+                    Poll::Ready(Err(err))
```

Comment:
Why should we not yield `Poll::Ready(Ok(()))` here in the case we did read some data in a previous iteration of the loop?

Reply:
> For tokio `AsyncRead`, even if an error is returned, the user can ignore the error and process data that has been read.

They technically can, but I don't think most will do so. The [`AsyncRead`](https://docs.rs/tokio/1.12.0/tokio/io/trait.AsyncRead.html) docs imply that implementations won't/shouldn't:

> On success, returns `Poll::Ready(Ok(()))` and places data in the unfilled portion of `buf`.

[`std::io::Read::read()`](https://doc.rust-lang.org/nightly/std/io/trait.Read.html#tymethod.read) (which is referenced in the `AsyncRead` docs) goes even further:

> If an error is returned then it must be guaranteed that no bytes were read.

I think approach #2 is reasonable, but am curious about why you prefer it (and/or why it wasn't the case previously).

Reply:
Okay, I think that makes sense.